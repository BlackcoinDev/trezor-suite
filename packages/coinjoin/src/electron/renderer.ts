/* eslint-disable max-classes-per-file */

// Replacement for @trezor/coinjoin index used in @suite-build/configs/desktop.webpack.config
// original CoinjoinClient and CoinjoinBackend classes are replaced by electron ipc implementation (see ./preloader)

export * as constants from '../constants';

export class CoinjoinClient {
    constructor(settings: any) {
        // @ts-expect-error
        const api = window.CoinjoinIpcChannel;
        if (!api) {
            throw new Error('Coinjoin ipc api not found');
        }
        return api.createClientInstance(settings);
    }
}

export class CoinjoinBackend {
    constructor(settings: any) {
        // @ts-expect-error
        const api = window.CoinjoinIpcChannel;
        if (!api) {
            throw new Error('Coinjoin ipc api not found');
        }
        return api.createBackendInstance(settings);
    }
}
