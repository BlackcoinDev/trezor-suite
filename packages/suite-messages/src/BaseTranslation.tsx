import React from 'react';
import { FormattedMessage, MessageDescriptor } from 'react-intl';
import { MessagesObject } from './messages';

// Use this type for project specific Translation components props
export type TranslationProps<TMessageId extends string> = {
    isNested?: boolean;
} & ExtendedMessageDescriptor<TMessageId>;

export type BaseTranslationProps<TMessageId extends string> = {
    isReactNative?: boolean; // used to render span on the web, plain node for react-native
    translationTooltip?: React.ComponentType<TooltipProps<TMessageId>>;
    messages: MessagesObject<TMessageId>;
} & TranslationProps<TMessageId>;

export interface TooltipProps<TMessageId extends string> {
    isNested?: boolean;
    messageId: TMessageId;
    children: any;
}

export type MessageValuesValue = string | number | boolean | Date | null | undefined;

// Add MessageDescriptor type to values entry
type FormatXMLElementFn = (...args: any[]) => string | object;
export interface ExtendedMessageDescriptor<TMessageId extends string> extends MessageDescriptor {
    id: TMessageId;
    values?: {
        [key: string]:
            | MessageValuesValue
            | React.ReactElement
            | ExtendedMessageDescriptor<TMessageId>
            | FormatXMLElementFn;
    };
}

const isMsgType = <TMessageId extends string>(
    props: BaseTranslationProps<TMessageId> | React.ReactNode,
): props is BaseTranslationProps<TMessageId> =>
    typeof props === 'object' &&
    props !== null &&
    (props as BaseTranslationProps<TMessageId>).id !== undefined;

const RenderChildrenComponent = ({ children }: any) => children;

export const BaseTranslation = <TMessageId extends string>(
    props: BaseTranslationProps<TMessageId>,
) => {
    // Tooltip is only used in suite-landing-page, in case landing will be moved to eshop this could be removed
    const TooltipComponent = props.translationTooltip ?? RenderChildrenComponent;
    const values: Record<
        string,
        MessageValuesValue | React.ReactNode | ExtendedMessageDescriptor<TMessageId>
    > = {};
    // message passed via props (id, defaultMessage, values)
    Object.keys(props.values || []).forEach(key => {
        // Iterates through all values. The entry may also contain a MessageDescriptor.
        // If so, Renders MessageDescriptor by passing it to `BaseTranslation` component
        const maybeMsg = props.values![key];
        values[key] = isMsgType(maybeMsg) ? (
            <BaseTranslation<TMessageId>
                {...maybeMsg}
                translationTooltip={props.translationTooltip}
                isNested
                messages={props.messages}
            />
        ) : (
            maybeMsg
        );
    });

    // prevent runtime errors
    if (
        !props.defaultMessage &&
        Object.prototype.hasOwnProperty.call(props, 'id') &&
        !props.messages[props.id]
    ) {
        return <>{`Unknown translation id: ${props.id}`}</>;
    }

    const defaultTagName = props.isReactNative || props.isNested ? undefined : 'span';

    return (
        <TooltipComponent isNested={props.isNested} messageId={props.id}>
            <FormattedMessage
                id={props.id}
                tagName={defaultTagName}
                defaultMessage={props.defaultMessage || props.messages[props.id].defaultMessage}
                // pass undefined to a 'values' prop in case of an empty values object
                values={Object.keys(values).length === 0 ? undefined : values}
            />
        </TooltipComponent>
    );
};
