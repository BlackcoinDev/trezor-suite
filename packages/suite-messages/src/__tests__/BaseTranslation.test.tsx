import React from 'react';

import renderer from 'react-test-renderer';
import { IntlProvider } from 'react-intl';
import { BaseTranslation, TranslationProps } from '../BaseTranslation';

const createComponentWithIntl = (children: React.ReactChild) =>
    renderer.create(<IntlProvider locale="en">{children}</IntlProvider>);

const messages = {
    TR_HELLO: {
        id: 'TR_HELLO',
        defaultMessage: 'Hello',
    },
    TR_HELLO_NAME: {
        id: 'TR_HELLO_NAME',
        defaultMessage: 'Hello {TR_NAME}, {TR_AGE}',
    },
    TR_NAME: {
        id: 'TR_NAME',
        defaultMessage: 'Name: {name}',
    },
    TR_AGE: {
        id: 'TR_NAME',
        defaultMessage: 'Age: {age}',
    },
} as const;

type MessageKey = keyof typeof messages;

const Translation = (props: TranslationProps<MessageKey>) => (
    <BaseTranslation messages={messages} {...props} />
);

describe('BaseTranslation component', () => {
    test('with id, defaultMessage props', () => {
        const component = createComponentWithIntl(<Translation id="TR_HELLO" />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('with message that holds string in value (passed via props)', () => {
        const component = createComponentWithIntl(
            <Translation {...messages.TR_NAME} values={{ name: 'John' }} />,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('with message that holds another message object in values (passed via props)', () => {
        const component = createComponentWithIntl(
            <Translation
                {...messages.TR_HELLO_NAME}
                values={{
                    TR_NAME: { ...messages.TR_NAME, values: { name: 'John' } },
                    TR_AGE: 100,
                }}
            />,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('with message that holds another in values (passed via props)', () => {
        const component = createComponentWithIntl(
            <Translation
                {...messages.TR_HELLO_NAME}
                values={{
                    TR_NAME: <Translation {...messages.TR_NAME} values={{ name: 'John' }} />,
                    TR_AGE: 100,
                }}
            />,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
