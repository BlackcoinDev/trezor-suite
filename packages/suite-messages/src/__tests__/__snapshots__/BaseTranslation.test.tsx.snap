// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTranslation component with id, defaultMessage props 1`] = `
<span>
  Hello
</span>
`;

exports[`BaseTranslation component with message that holds another in values (passed via props) 1`] = `
<span>
  Hello 
  <span>
    Name: John
  </span>
  , 100
</span>
`;

exports[`BaseTranslation component with message that holds another message object in values (passed via props) 1`] = `
<span>
  Hello 
  Name: John
  , 100
</span>
`;

exports[`BaseTranslation component with message that holds string in value (passed via props) 1`] = `
<span>
  Name: John
</span>
`;
